### ISSUE TITLE
Add repo alias support to requirements.yaml

### ISSUE THREAD

**Issue by micahhausler** [kubernetes/helm#1733]
A pain point with `requirements.yaml` files in charts is that `repository` is a required field in each dependency. 

For example, say I have a chart called `webapp-1.0.0` with a required chart `base-chart-0.1.0`.  In order to have a simple CI process around dependent charts, I _must_ already have the `base-chart`‘s tarball built, served, and fetched _from the correct repository_ in `webapp`‘s `/charts/` directory.

`webapp`'s `requirements.yaml`
```yaml
dependencies:
- name: base-chart
  version: 0.1.0
  repository: http://my-production-chart-repo.com
```

I'm not sure that there is a great workaround, other than making sure that the example `base-chart` updates and CI builds are separate from any other chart that requires it.

**Comment by brandoncole** [comment]
There's two concerns for me related to this:

1. The canonical URL for our helm repository is under flux, it could be `http://helm.company.com` or `http://product.company.com/charts`.  To further complicate things, we have Alpha and Beta environments so it could also be `http://helm.alpha.company.com` or `http://helm.beta.company.com`.  So we'd prefer not to hardcode a URL for the `repository` field in all our charts, if the repository is the local repository it'd be nice if it were something like:

```yaml
dependencies:
  - name: postgres
    version: 0.2.0
    repository: local
  - name: provisioner
    version: 0.2.0
    repository: local
```

2. Alternatively, and perhaps even better, it'd be nice if we could refer to the repo alias instead.  For example:

```bash
$ helm repo list
NAME     	URL
stable   	http://storage.googleapis.com/kubernetes-charts
local    	http://localhost:8879/charts
stablerepo	http://some.environment.specific.url.io/charts
```

```yaml
dependencies:
  - name: postgres
    version: 0.2.0
    repository: stablerepo
  - name: provisioner
    version: 0.2.0
    repository: stablerepo
```

**Comment by jascott1** [comment]
file:// could also be used to satisfy URL requirement but it does tie the dependencies to local system.

**Comment by technosophos** [comment]
For the sake of CI systems, I would be okay with adding a convention with something like:

```
repository: "alias:local" # match to the name "local" and fetch by URL
...
repository: "file://../foo/bar" # match relative to the local filesystem
```

But both of these significantly damage the portability of the `requirements.yaml` file since they bind resolution to the arbitrary configuration of a local system. So I would then recommend that most people not use them, and encourage the Kubernetes Charts repo to explicitly disallow using either of these forms.

**Comment by jascott1** [comment]
@technosophos Sounds reasonable. Kolla has aspirations to be in charts/stable though so that would preclude us from using this. If we could us file://, we think we could remove the script that does dep build and dep up on all the charts, which would be nice since it it needs to be run for each install/upgrade after local changes are made. @kfox1111 can probably add some insight into the usefulness of local paths for kolla as well.

**Comment by jascott1** [comment]
Should also add that @qwangrepos is working on proposal for this.

**Comment by kfox1111** [comment]
the use case is covered in #1639

Do you need more info?

**Comment by kfox1111** [comment]
@technosophos I disagree with the portability assertion.

The current url scheme makes some assumptions too that damage portability:
1, that you can access the same url the dev used (enterprise firewall issues/mirrors need to be used)
2, that if using localhost:<someport>, that your <someport> matches the devs's <someport>

And semi related, requires a version field. Which makes sense for remote server pulls. Works well when your deps are released independently of yours. More painful if you have a lot of deps you yourself maintain. I'd rather keep all the deps together in the same git repo, tag them in git, and build them all, one shot. ensures exactly the right version is used, so doesn't really require version numbers on the deps.

**Comment by technosophos** [comment]
The `file:///` support was added in 2.2.0.

If anyone wants to push forward the `alias:` support, I can leave this open. Otherwise, I'll close it out.

**Comment by technosophos** [comment]
@qwangrepos has asked to claim this one for 2.3.0, and will provide a spec.

**Comment by technosophos** [comment]
Unless the PR for this is already in progress, can we bump this to 2.4.0 @qwangrepos ?

**Comment by qwangrepos** [comment]
I'm working on it now but feel free to bump it to 2.4.0.

**Comment by thomastaylor312** [comment]
@qwangrepos Looks like @technosophos added the PR for this (#2174) to 2.3.0 after he made this comment, so I am assuming we are ok to put it in 2.3

**Comment by technosophos** [comment]
The relevant patch got merged as #2174. Marking it complete. Thanks again!
